lib/ProductOpener/CRM.pm

```diff
@@ -0,0 +1,257 @@
+	$log->debug("create_company", {company_id => $company_id}) if $log->is_debug();
+	return $company_id;
+}
+
+
+
+=head2 create_contact ($user_ref)
+
+Creates a new contact in odoo from a user
+
+=head3 Arguments
+
+=head4 $org_ref
+
+=head3 Return values
+
+the id of the created contact
+
+=cut
+
+sub create_contact ($user_ref) {
+	my $contact = {
+		name => 			$user_ref->{name}, 
+		x_off_username => 	$user_ref->{userid},
+		email => 			$user_ref->{email},
+		phone => 			$user_ref->{phone}, 
+		category_id => 		[$odoo_tags{Producter}],
+	};
+	
+	# find country code id in odoo
+	my $user_country_code = uc $country_codes_reverse{$user_ref->{country}} || 'EN';
+	my $country_id = odoo('res.country', 'search_read', [[['code', '=', $user_country_code]]], { 'fields' => ['code', 'id']});
+	$contact->{country_id} = $country_id->[0]{id};
+
+	# find spoken language's code in Odoo
+	my $odoo_lang = odoo('res.lang', 'search_read', [[['iso_code', '=', $user_ref->{preferred_language}]]], { 'fields' => ['code']});
+	my $found_lang = $odoo_lang->[0];
+	$contact->{lang} = $found_lang->{code} // 'en_US'; # default to english
+	if(defined $found_lang->{id}) {
+		$contact->{x_off_languages} = [$found_lang->{id}];
+	}
+
+	# create new contact with associated organization
+	my $contact_id = odoo('res.partner', 'create', [{%$contact}]);
+	$log->debug("create_contact", {contact_id => $contact_id}) if $log->is_debug();
+	return $contact_id;
+}
+
+=head2 add_contact_to_company ($org_ref)
+
+Add a contact to a company in odoo
+
+=head3 Arguments
+
+=head4 $crm_user_id
+
+=head4 $crm_org_id
+
+=head3 Return values
+
+1 if success, undef otherwise
+
+=cut
+
+sub add_contact_to_company($crm_user_id, $crm_org_id) {
+	# set contact in children of company
+	# 4 means add an existing record
+	odoo('res.partner', 'write', [[$crm_org_id], {child_ids => [[4, $crm_user_id]]}]);
+	# set company of the contact
+	my $result = odoo('res.partner', 'write', [[$crm_user_id], {parent_id => $crm_org_id}]);
+	$log->debug("add_contact_to_company", {company_id => $crm_org_id, contact_id => $crm_user_id}) if $log->is_debug();
+	return $result
+}
+
+=head2 create_opportunity ($user_ref, $org_ref)
+
+create an opportunity attached to a user 
+and named after the organization
+
+=head3 Arguments
+
+=head4 $user_ref
+
+=head4 $org_ref
+
+=head3 Return values
+
+the id of the created opportunity
+
+=cut
+
+sub create_opportunity ($user_ref, $org_ref) {
+	my $opportunity_id = odoo('crm.lead', 'create', [{name => "$org_ref->{name} - new org", partner_id => $user_ref->{crm_user_id}}]);
+	$log->debug("create_opportunity", {opportunity_id => $opportunity_id}) if $log->is_debug();
+	return $opportunity_id;
+}
+
+
+=head2 odoo (@params)
+
+call odoo api with the given parameters
+
+=head3 Return values
+
+the response or undef if an error occured
+
+=cut
+
+sub odoo(@params) {
+	if(not defined $xmlrpc) {
+		my $api_url = $ENV{ODOO_CRM_URL} . '//xmlrpc/2/';
```

Comment:
We don't really use `$ENV` for dynamic configuration like this, outside of [`Config2_docker.pm`](https://github.com/openfoodfacts/openfoodfacts-server/blob/ece6f77e9333e5e35f727ac71f7b1f272f5f7cbb/lib/ProductOpener/Config2_docker.pm). The "proper" way to add optional odoo configuration is probably adding it to `Config.pm` and to `Config2.pm`